### PR TITLE
Feat: Add basic LoRA training support

### PR DESCRIPTION
**Pull request by yoland68**
For more details: https://github.com/Comfy-Org/rfcs/pull/26

### CODE REVIEW

**Comment by yoland68** [comment]
image

**Comment by mcDandy** [comment]
Thank you for an option for style transfer.

I think it might be a good idea to add a progress bar as is on sampler nodes. It takes approx 4x time of interference so seeing a visual progress of how it is going is a great quaility of life improvement.

**Comment by yoland68** [comment]
> Thank you for an option for style transfer.
> 
> I think it might be a good idea to add a progress bar as is on sampler nodes. It takes approx 4x time of interference so seeing a visual progress of how it is going is a great quaility of life improvement.

Yup, having progress would be great, the annoying thing right now with progress is that it's a hack. I m hoping we can support an official progress API for this. Will prob add the hack for now

File: comfy/ldm/modules/attention.py [hunk 1]
```diff
@@ -673,7 +673,7 @@ def forward(self, x, context=None, transformer_options={}):
             for p in patch:
                 n = p(n, extra_options)
 
-        x += n
+        x = n + x
```
Comment (huchenlei) [inline comfy/ldm/modules/attention.py:753]: Can you explain why this is necessary? In my mind the expression here should be equivalent.

File: comfy_extras/nodes_train.py [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+
+    RETURN_TYPES = ("IMAGE",)
+    FUNCTION = "load_images"
+    CATEGORY = "loaders"
+    EXPERIMENTAL = True
+    DESCRIPTION = "Loads a batch of images from a directory for training."
+
+    def load_images(self, folder, resize_method):
+        sub_input_dir = os.path.join(folder_paths.get_input_directory(), folder)
+        valid_extensions = [".png", ".jpg", ".jpeg", ".webp"]
```
Comment (huchenlei) [inline comfy_extras/nodes_train.py:211]: Access allow pillow supported formats with following code:
```
from PIL import Image
print(Image.registered_extensions())  # Shows all registered extensions and their formats
```

```json
{'.blp': 'BLP', '.bmp': 'BMP', '.dib': 'DIB', '.bufr': 'BUFR', '.cur': 'CUR', '.pcx': 'PCX', '.dcx': 'DCX', '.dds': 'DDS', '.ps': 'EPS', '.eps': 'EPS', '.fit': 'FITS', '.fits': 'FITS', '.fli': 'FLI', '.flc': 'FLI', '.ftc': 'FTEX', '.ftu': 'FTEX', '.gbr': 'GBR', '.gif': 'GIF', '.grib': 'GRIB', '.h5': 'HDF5', '.hdf': 'HDF5', '.png': 'PNG', '.apng': 'PNG', '.jp2': 'JPEG2000', '.j2k': 'JPEG2000', '.jpc': 'JPEG2000', '.jpf': 'JPEG2000', '.jpx': 'JPEG2000', '.j2c': 'JPEG2000', '.icns': 'ICNS', '.ico': 'ICO', '.im': 'IM', '.iim': 'IPTC', '.jfif': 'JPEG', '.jpe': 'JPEG', '.jpg': 'JPEG', '.jpeg': 'JPEG', '.mpg': 'MPEG', '.mpeg': 'MPEG', '.tif': 'TIFF', '.tiff': 'TIFF', '.mpo': 'MPO', '.msp': 'MSP', '.palm': 'PALM', '.pcd': 'PCD', '.pdf': 'PDF', '.pxr': 'PIXAR', '.pbm': 'PPM', '.pgm': 'PPM', '.ppm': 'PPM', '.pnm': 'PPM', '.pfm': 'PPM', '.psd': 'PSD', '.qoi': 'QOI', '.bw': 'SGI', '.rgb': 'SGI', '.rgba': 'SGI', '.sgi': 'SGI', '.ras': 'SUN', '.tga': 'TGA', '.icb': 'TGA', '.vda': 'TGA', '.vst': 'TGA', '.webp': 'WEBP', '.wmf': 'WMF', '.emf': 'WMF', '.xbm': 'XBM', '.xpm': 'XPM'}
```

File: execution.py [hunk 3]
```diff
@@ -573,7 +584,7 @@ def validate_inputs(prompt, item, validated):
         val = inputs[x]
         info = (type_input, extra_info)
         if isinstance(val, list):
-            if len(val) != 2:
+            if len(val) != 2 and not extra_info.get("allow_batch", False):
```
Comment (huchenlei) [inline execution.py]: Can we split this fix to a separate PR?

**Review by huchenlei — CHANGES_REQUESTED** [review]
Content LGTM.

I would like request split the execution change and addition of image set nodes to a separate PR for better understanding, and management (In case we want to revert certain part).

File: comfy/ldm/modules/attention.py (hunk 1 above)
Comment (yoland68) [reply comfy/ldm/modules/attention.py:753]: https://medium.com/@rizqinur2010/pytorch-in-place-operations-bfde2f171afa

Comment (yoland68) [reply comfy/ldm/modules/attention.py:753]: Torch in place operation can make gradient compute difficult

File: execution.py (hunk 3 above)
Comment (yoland68) [reply execution.py]: man, PR split should really be an LLM agent's work

File: comfy_extras/nodes_train.py (hunk 2 above)
Comment (yoland68) [reply comfy_extras/nodes_train.py:211]: Added a few more common format but skipped on all the random ones given unlikely users will add them.

Comment (christian-byrne) [reply comfy_extras/nodes_train.py:211]: You could use this with `["image"]` as the content type arg:

https://github.com/comfyanonymous/ComfyUI/blob/ca8efab79fa19bc9745b4f7346d38a49ba1b1b7c/comfy_extras/nodes_audio.py#L218-L219

File: comfy/ldm/modules/attention.py [hunk 4]
```diff
@@ -713,12 +713,12 @@ def forward(self, x, context=None, transformer_options={}):
...
         if self.is_res:
             x_skip = x
         x = self.ff(self.norm3(x))
         if self.is_res:
-            x += x_skip
+            x = x_skip + x
```
Comment (yewentao256) [inline comfy/ldm/modules/attention.py:798]: I am thinking if we could still use the inplace operator when it is not a leaf node.

Eg:

```py
import torch

def add_1_optimized(x: torch.Tensor) -> torch.Tensor:
    if x.is_leaf:
        x = x + 1
    else:
        x += 1
    return x

def add_1(x: torch.Tensor) -> torch.Tensor:
    return x + 1

import time

x = torch.randn(100, 200, 200, 200, requires_grad=True)
start = time.time()
y = add_1_optimized(x)
z = add_1_optimized(y)
assert id(y) == id(z)
g = add_1_optimized(z)
h = add_1_optimized(g)
i = add_1_optimized(h)
print('Time taken for add_1_optimized:', time.time() - start)

start = time.time()
i.sum().backward()
print('Time taken for add_1_optimized backward:', time.time() - start)

x = torch.randn(100, 200, 200, 200, requires_grad=True)
start = time.time()
y = add_1(x)
z = add_1(y)
g = add_1(z)
h = add_1(g)
i = add_1(h)
print('Time taken for add_1:', time.time() - start)
start = time.time()
i.sum().backward()
print('Time taken for add_1 backward:', time.time() - start)
```

Result:

```bash
Time taken for add_1_optimized: 1.166808843612671
Time taken for add_1_optimized backward: 0.3505527973175049
Time taken for add_1: 1.8314952850341797
Time taken for add_1 backward: 0.395704984664917
```
